File: go/internal/api/security.go
// Package api — security middleware (anti-DoS / response-header hardening).
//
// This file ships two chi-compatible middleware factories:
//
//   - SecurityHeaders() — sets Strict-Transport-Security,
//     Content-Security-Policy, and X-Content-Type-Options on every response.
//   - MaxBytesReader(limit) — bounds the REST request body at `limit` bytes
//     via a two-tier defense: Tier 1 (Content-Length pre-check) emits 413 +
//     Twilio JSON BEFORE reading the body; Tier 2 (http.MaxBytesReader body
//     wrap) catches chunked-encoded oversize bodies that lie about
//     Content-Length or omit it.
//
// Both factories are wired in api/server.go Mount() BEFORE BasicAuth so
// security headers are present on the 401 path and 413 oversize-body refusal
// fires before any credential check (anti-DoS — minimum work to reject).
//
// REST endpoints enforce a 64KB max body and emit security headers; a
// deliberate MaxBytesReader overflow returns 413 with the Twilio JSON
// error body.
package api

import (
	"net/http"
)

// SecurityHeaders returns chi-compatible middleware that sets the three
// locked security headers on every REST response (including 401/4xx error
// responses). Headers are written BEFORE next.ServeHTTP, so even if the next
// handler returns an error, the operator gets HSTS/CSP/nosniff.
//
// Locked header set:
//
//   - Strict-Transport-Security: max-age=63072000; includeSubDomains
//     (2-year HSTS, per OWASP cheatsheet; preload not asserted because the
//     bridge runs behind a K8s ingress that owns public TLS termination).
//   - Content-Security-Policy: default-src 'none'; frame-ancestors 'none'
//     (REST-only API; no HTML/script content; minimum-permissive policy).
//   - X-Content-Type-Options: nosniff (defense against MIME-type sniffing
//     attacks on JSON bodies).
//
// Used by Mount() in api/server.go BEFORE BasicAuth so security headers are
// present even on 401 unauthenticated responses (load-bearing ordering;
// pinned by TestSecurityHeaders_PresentOn401).
func SecurityHeaders() func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			h := w.Header()
			h.Set("Strict-Transport-Security", "max-age=63072000; includeSubDomains")
			h.Set("Content-Security-Policy", "default-src 'none'; frame-ancestors 'none'")
			h.Set("X-Content-Type-Options", "nosniff")
			next.ServeHTTP(w, r)
		})
	}
}

// MaxBytesReader returns chi-compatible middleware that bounds every REST
// request body at `limit` bytes.
//
// Two-tier defense — both tiers required:
//
//  1. **Tier 1 (Content-Length pre-check):** if r.ContentLength > limit,
//     emit a Twilio-shaped 413 JSON body immediately and return without
//     reading any body bytes. Rejects pre-declared oversize at the cheapest
//     possible point. Covers the canonical attack shape — curl,
//     twilio-python, twilio-go, twilio-node and every other legitimate
//     client all set Content-Length on POSTs.
//
//  2. **Tier 2 (http.MaxBytesReader body wrap):** wrap r.Body with
//     http.MaxBytesReader(w, r.Body, limit). Catches chunked-encoded
//     oversize bodies that lie about Content-Length (Content-Length: 100
//     with a 1MB chunked body) or omit Content-Length entirely
//     (Transfer-Encoding: chunked with no CL). The downstream handler
//     reading r.Body (e.g. modifyCallHandler's r.ParseForm()) receives a
//     *http.MaxBytesError once limit bytes are consumed; the existing
//     parseModifyOpts error path emits 400 invalid_params for that case.
//     Without Tier 2, an attacker trivially bypasses Tier 1 by setting
//     Content-Length: 0 + Transfer-Encoding: chunked + 1GB body.
//
// The 4xx-vs-413 asymmetry on chunked oversize is accepted — chunked
// oversize still gets refused with a 4xx and operator monitoring still
// surfaces the rejection via api_requests_total{status="4xx"}.
//
// Production cap is 64<<10 (65536 bytes) — sized at 16x the largest
// legitimate Twiml= body (4000 chars) plus headroom for status-callback-
// related form fields.
func MaxBytesReader(limit int64) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Tier 1: cheap Content-Length pre-check.
			//
			// Strict `>`: a request with Content-Length == limit is allowed
			// through. Tier 2's http.MaxBytesReader is inclusive (allows
			// exactly `limit` bytes; errors on the (limit+1)th read), so the
			// boundary stays consistent across the two tiers.
			if r.ContentLength > limit {
				ErrPayloadTooLarge().WriteJSON(w)
				return
			}
			// Tier 2: wrap r.Body so chunked-encoding / fraudulent-CL /
			// unknown-length requests are bounded. Downstream handler reading
			// r.Body receives *http.MaxBytesError once limit bytes are
			// consumed.
			r.Body = http.MaxBytesReader(w, r.Body, limit)
			next.ServeHTTP(w, r)
		})
	}
}

// Compile-time assertions that the factories satisfy chi's middleware
// signature (which is structurally `func(http.Handler) http.Handler` —
// chi.Router.Use accepts this via Go's interface assignability without a
// chi-typed import). The mount-site assertion happens organically when
// api/server.go calls `r.Use(SecurityHeaders())` and
// `r.Use(MaxBytesReader(64<<10))` on a `chi.Router` — that's the load-bearing
// compatibility check. These assertions are belt-and-suspenders to catch
// signature drift earlier than mount time.
var (
	_ func(http.Handler) http.Handler = SecurityHeaders()
	_ func(http.Handler) http.Handler = MaxBytesReader(64 << 10)
)

File: node/src/api/security.ts
/**
 * Security middleware for the Twilio-compatible REST control plane
 * (response-header hardening + anti-DoS body bounding).
 *
 * Ported from the Go chi middleware internal/api/security.go. The Go version
 * ships two `func(http.Handler) http.Handler` factories; this project serves via
 * raw node:http (http.createServer, NOT express/chi), so each piece is modelled
 * as a plain function over (req, res):
 *
 *   - securityHeaders(res)              — side-effecting; sets the three locked
 *                                          headers on every response.
 *   - maxBytes(req, res, limit): boolean — Tier-1 Content-Length pre-check;
 *                                          returns false after writing 413.
 *   - readFormBody(req, limit)          — Tier-2 stream-bounded body reader that
 *                                          parses application/x-www-form-urlencoded.
 *
 * Compose in this fixed order so headers land on the 401 path and oversize
 * bodies are rejected before any credential work:
 *
 *   securityHeaders(res)  →  maxBytes(req, res)  →  basicAuth(req, res, …)  →  handler
 */

import type { IncomingMessage, ServerResponse } from 'node:http';

import { ErrPayloadTooLarge } from './errors.js';

/** Production REST body cap — 64KiB (64 << 10). Sized at 16x the largest
 * legitimate Twiml= body (4000 chars) plus headroom for status-callback fields. */
export const DEFAULT_MAX_BYTES = 65_536;

/** Thrown by readFormBody when the body stream exceeds the configured cap.
 * Distinguishable via instanceof so callers can emit a 413 / Twilio 21617. */
export class PayloadTooLargeError extends Error {
  readonly limit: number;
  constructor(limit: number) {
    super(`request body exceeds ${limit} byte limit`);
    this.name = 'PayloadTooLargeError';
    this.limit = limit;
  }
}

/**
 * Set the three locked security headers on EVERY response — including 401/413
 * error responses. Call before any stage that may write a response so the
 * headers are buffered ahead of writeHead.
 *
 * Locked set (values byte-identical to the Go middleware):
 *   - Strict-Transport-Security: max-age=63072000; includeSubDomains  (2-year HSTS)
 *   - Content-Security-Policy: default-src 'none'; frame-ancestors 'none'
 *   - X-Content-Type-Options: nosniff
 */
export function securityHeaders(res: ServerResponse): void {
  res.setHeader('Strict-Transport-Security', 'max-age=63072000; includeSubDomains');
  res.setHeader('Content-Security-Policy', "default-src 'none'; frame-ancestors 'none'");
  res.setHeader('X-Content-Type-Options', 'nosniff');
}

/**
 * Parse a Content-Length header into a non-negative integer, or undefined when
 * the header is absent or unparseable (chunked / unknown length). node:http may
 * surface the header as a string or — never, in practice — an array; both are
 * handled defensively.
 */
function parseContentLength(req: IncomingMessage): number | undefined {
  const raw = req.headers['content-length'];
  const value = Array.isArray(raw) ? raw[0] : raw;
  if (value === undefined) {
    return undefined;
  }
  const n = Number(value);
  if (!Number.isInteger(n) || n < 0) {
    return undefined;
  }
  return n;
}

/**
 * Tier-1 Content-Length pre-check.
 *
 * If the declared Content-Length STRICTLY exceeds `limit`, write a Twilio-shaped
 * 413 immediately (before reading any body bytes) and return false. Otherwise
 * return true. A request with Content-Length == limit is allowed through —
 * readFormBody's Tier-2 cap is inclusive (allows exactly `limit` bytes), so the
 * boundary stays consistent across the two tiers.
 *
 * Tier 2 (the stream cap) lives in readFormBody so chunked / fraudulent-CL /
 * unknown-length bodies are bounded when actually consumed.
 */
export function maxBytes(req: IncomingMessage, res: ServerResponse, limit: number = DEFAULT_MAX_BYTES): boolean {
  const contentLength = parseContentLength(req);
  if (contentLength !== undefined && contentLength > limit) {
    ErrPayloadTooLarge().writeJSON(res);
    return false;
  }
  return true;
}

/**
 * Read the full request body, enforcing a hard byte cap on the actual stream
 * (Tier 2). Rejects with PayloadTooLargeError as soon as consumption STRICTLY
 * exceeds `limit` — i.e. exactly `limit` bytes are allowed; the (limit+1)th
 * byte trips the cap. This catches chunked-encoded oversize bodies that lie
 * about Content-Length or omit it, which Tier-1 (maxBytes) cannot see.
 */
export function readBody(req: IncomingMessage, limit: number = DEFAULT_MAX_BYTES): Promise<Buffer> {
  return new Promise<Buffer>((resolve, reject) => {
    const chunks: Buffer[] = [];
    let total = 0;
    let settled = false;

    const cleanup = (): void => {
      req.removeListener('data', onData);
      req.removeListener('end', onEnd);
      req.removeListener('error', onError);
      req.removeListener('aborted', onAborted);
    };
    const fail = (err: Error): void => {
      if (settled) {
        return;
      }
      settled = true;
      cleanup();
      // Stop pulling further bytes from an oversize / failed stream.
      req.destroy();
      reject(err);
    };
    const succeed = (buf: Buffer): void => {
      if (settled) {
        return;
      }
      settled = true;
      cleanup();
      resolve(buf);
    };

    const onData = (chunk: Buffer): void => {
      total += chunk.length;
      if (total > limit) {
        fail(new PayloadTooLargeError(limit));
        return;
      }
      chunks.push(chunk);
    };
    const onEnd = (): void => {
      succeed(Buffer.concat(chunks));
    };
    const onError = (err: Error): void => {
      fail(err);
    };
    const onAborted = (): void => {
      fail(new Error('request aborted'));
    };

    req.on('data', onData);
    req.on('end', onEnd);
    req.on('error', onError);
    req.on('aborted', onAborted);
  });
}

/**
 * Read and parse an application/x-www-form-urlencoded request body, enforcing
 * the Tier-2 byte cap. Used by handlers such as modifyCallHandler.
 *
 * - Oversize → rejects with PayloadTooLargeError (caller emits 413 / 21617).
 * - Stream errors → reject verbatim.
 * - Parsing is delegated to URLSearchParams, which never throws on malformed
 *   input; bare/garbage tokens simply produce empty-value or dropped pairs.
 *   Callers validate required fields downstream and surface 400 invalid_params.
 */
export async function readFormBody(req: IncomingMessage, limit: number = DEFAULT_MAX_BYTES): Promise<URLSearchParams> {
  const buf = await readBody(req, limit);
  return new URLSearchParams(buf.toString('utf8'));
}
